### PR TITLE
Out of Memory Crash: executorService

### PR DESCRIPTION
**Pull request by zafirskthelifehacker**
2022-03-17 14:09:28.211 30068-13841/com.sharinpix.SharinPix A/inpix.SharinPi: thread.cc:4192] Unable to create protected region in stack for implicit overflow check. Reason: Out of memory size:  4096

### CODE REVIEW

File: src/android/FileTransferBackground.java [hunk 1]
```diff
@@ -319,7 +328,7 @@ private void addUpload(JSONObject jsonPayload) {
     private void startUpload(final String uploadId, final Data payload) {
         Log.d(TAG, "startUpload: Starting work via work manager");
 
-        OneTimeWorkRequest workRequest = new OneTimeWorkRequest.Builder(UploadTask.class)
+        OneTimeWorkRequest.Builder workRequestBuilder = new OneTimeWorkRequest.Builder(UploadTask.class)
                 .setConstraints(new Constraints.Builder()
                         .setRequiredNetworkType(NetworkType.CONNECTED)
                         .build()
```
Comment (zafirskthelifehacker) [inline src/android/FileTransferBackground.java:334]: Samsung Devices: Uploading 300 images at a time blocks queue of uploads
On Pixel Devices works Fine
Removing Constraints in the WorkRequest helps to allow upload the image normally

File: src/android/FileTransferBackground.java [hunk 2]
```diff
@@ -181,54 +190,64 @@ private void initManager(String options, final CallbackContext callbackContext)
...
-            handleAck(ack.getOutputData());
-        }
+        new Thread() {
+            @Override
+            public void run() {
+                for (UploadEvent ack : uploadEvents) {
+                    handleAck(ack.getOutputData());
+                }
+                try {
+                    sleep(250);
```
Comment (azharbeebeejaun) [inline src/android/FileTransferBackground.java:200]: Or use https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html

File: src/android/FileTransferBackground.java [hunk 3]
```diff
@@ -181,54 +190,64 @@ private void initManager(String options, final CallbackContext callbackContext)
...
+
+                                    Log.d(TAG, "initManager: " + info.getId() + " (" + info.getState() + ") Progress: " + progress);
+                                    sendProgress(id, progress);
+                                }
+                                break;
+                            case CANCELLED:
+                            case BLOCKED:
+                            case ENQUEUED:
+                            case SUCCEEDED:
+                                completedTasks = completedTasks + 1;
```
Comment (azharbeebeejaun) [inline src/android/FileTransferBackground.java:230]:
```suggestion
                                completedTasks++;
```
?

File: src/android/FileTransferBackground.java [hunk 4]
```diff
@@ -181,54 +190,64 @@ private void initManager(String options, final CallbackContext callbackContext)
...
-                            switch (info.getState()) {
-                                // If the upload in not finished, publish its progress
-                                case RUNNING:
-                                    if (info.getProgress() != Data.EMPTY) {
-                                        String id = info.getProgress().getString(UploadTask.KEY_PROGRESS_ID);
-                                        int progress = info.getProgress().getInt(UploadTask.KEY_PROGRESS_PERCENT, 0);
-
-                                        Log.d(TAG, "initManager: " + info.getId() + " (" + info.getState() + ") Progress: " + progress);
-                                        sendProgress(id, progress);
+        // Listen for upload progress
```
Comment (azharbeebeejaun) [inline src/android/FileTransferBackground.java:209]: Fix indent?

File: src/android/FileTransferBackground.java [hunk 5]
```diff
@@ -517,6 +541,7 @@ public static String fetchCurrentTag(Context context) {
                 }
             }
         }
+        AckDatabase.closeInstance();
```
Comment (azharbeebeejaun) [inline src/android/FileTransferBackground.java:544]: Pas sûr si ça devrait être là ? On va ouvrir la connection juste après non car c'est à peu près quand `fetchCurrentTag` est appelée (au début).

File: src/android/FileTransferBackground.java [hunk 6]
```diff
@@ -181,9 +195,14 @@ private void initManager(String options, final CallbackContext callbackContext)
...
                 .getAll();
 
-        for (UploadEvent ack : uploadEvents) {
-            handleAck(ack.getOutputData());
-        }
+        scheduledExecutorService.schedule(new Runnable() {
+            @Override
+            public void run() {
+                for (UploadEvent ack : uploadEvents) {
+                    handleAck(ack.getOutputData());
```
Comment (azharbeebeejaun) [inline src/android/FileTransferBackground.java:202]: `schedule` ça plutôt

File: src/android/FileTransferBackground.java [hunk 7]
```diff
@@ -445,9 +469,11 @@ private void cleanupUpload(final String uploadId) {
     }
 
     public void destroy() {
+        AckDatabase.closeInstance();
```
Comment (azharbeebeejaun) [inline src/android/FileTransferBackground.java:472]: Pas sûr qu'on le veut ici... Je suis pour le mettre en commentaire jusqu'à ce qu'on trouve le bon endroit.

File: src/android/AckDatabase.java [hunk 8]
```diff
@@ -23,5 +23,10 @@ public static AckDatabase getInstance(final Context context) {
         return instance;
     }
 
+    public static void closeInstance() {
+        instance.close();
```
Comment (azharbeebeejaun) [inline src/android/AckDatabase.java:27]:
```suggestion
        // not called for now
        instance.close();
```

File: src/android/FileTransferBackground.java [hunk 9]
```diff
@@ -52,6 +56,9 @@
     private static String currentTag;
     private static long currentTagFetchedAt;
 
+    private ExecutorService executorService = null;
+    private ScheduledExecutorService scheduledExecutorService = null;
```
Comment (azharbeebeejaun) [inline src/android/FileTransferBackground.java:60]: Can use only 1 and set delay 0 for some. Initial events can be sent at fixed intervals.

File: src/android/FileTransferBackground.java [hunk 10]
```diff
@@ -181,8 +190,11 @@ private void initManager(String options, final CallbackContext callbackContext)
                 .uploadEventDao()
                 .getAll();
 
+        int acknowledgeScheduler = 0;
         for (UploadEvent ack : uploadEvents) {
-            handleAck(ack.getOutputData());
+            executorService.schedule(() -> {
+                handleAck(ack.getOutputData());
+            }, acknowledgeScheduler += 250, TimeUnit.MILLISECONDS);
```
Comment (azharbeebeejaun) [inline src/android/FileTransferBackground.java:197]:
```suggestion
            }, acknowledgeScheduler += 200, TimeUnit.MILLISECONDS);
```

File: src/android/FileTransferBackground.java [hunk 11]
```diff
@@ -181,8 +190,11 @@ private void initManager(String options, final CallbackContext callbackContext)
                 .uploadEventDao()
                 .getAll();
 
+        int acknowledgeScheduler = 0;
```
Comment (azharbeebeejaun) [inline src/android/FileTransferBackground.java:193]:
```suggestion
        int ackDelay = 0;
```

**Review by ombr — APPROVED** [review]

**Review by dinitri-ragoo — APPROVED** [review]
the greater the number of images , the slower it is for a single upload. however all images have been successfully uploaded

**Review by azharbeebeejaun — APPROVED** [review]

File: src/android/FileTransferBackground.java [hunk 12]
```diff
@@ -116,7 +121,11 @@ private void sendError(final String id, final String reason, boolean userCancele
 
     @Override
     public boolean execute(String action, JSONArray args, final CallbackContext callbackContext) {
-        cordova.getThreadPool().execute(() -> {
+        if(executorService == null) {
+            executorService = Executors.newScheduledThreadPool(4);
```
Comment (azharbeebeejaun) [inline src/android/FileTransferBackground.java:125]: https://github.com/apache/cordova-android/blob/2a84d7c44ddf5eb169f34a2448915a53d0a30fcd/framework/src/org/apache/cordova/CordovaInterfaceImpl.java#L56
https://stackoverflow.com/questions/949355/executors-newcachedthreadpool-versus-executors-newfixedthreadpool

**Pull request merged**